sound/soc/sof/core.c

```diff
@@ -496,13 +516,26 @@ int snd_sof_device_probe(struct device *dev, struct snd_sof_pdata *plat_data)
 		return 0;
 	}
 
-	return sof_probe_continue(sdev);
+	ret = sof_probe_continue(sdev);
+	if (ret < 0)
+		return ret;
+
+	/* enable runtime_pm */
+	pm_runtime_set_autosuspend_delay(&pdev->dev, SND_SOF_SUSPEND_DELAY_MS);
+	pm_runtime_use_autosuspend(&pdev->dev);
+
+	pm_runtime_allow(&pdev->dev);
+
+	/* mark last_busy for pm_runtime to make sure not suspend immediately */
+	pm_runtime_mark_last_busy(&pdev->dev);
+
+	return ret;
 }
 EXPORT_SYMBOL(snd_sof_device_probe);
 
-int snd_sof_device_remove(struct device *dev)
+int snd_sof_device_remove(struct platform_device *pdev)
 {
-	struct snd_sof_dev *sdev = dev_get_drvdata(dev);
+	struct snd_sof_dev *sdev = dev_get_drvdata(pdev->dev.parent);
```

Comment:
I don't like this, you are mixing layers. We should not have any confusion between DSP driver and parent.